src/operations/client_bulk_write/command_builder.ts

```diff
@@ -62,32 +67,148 @@ export class ClientBulkWriteCommandBuilder {
   /**
    * Build the bulk write commands from the models.
    */
-  buildCommands(): ClientBulkWriteCommand[] {
+  buildCommands(maxMessageSizeBytes: number, maxWriteBatchSize: number): ClientBulkWriteCommand[] {
     // Iterate the models to build the ops and nsInfo fields.
-    const operations = [];
+    // We need to do this in a loop which creates one command each up
+    // to the max bson size or max message size.
+    const commands: ClientBulkWriteCommand[] = [];
+    let currentCommandLength = 0;
     let currentNamespaceIndex = 0;
+    let currentCommand: ClientBulkWriteCommand = this.baseCommand();
     const namespaces = new Map<string, number>();
+
     for (const model of this.models) {
       const ns = model.namespace;
       const index = namespaces.get(ns);
+
+      /**
+       * Convenience function for resetting everything when a new batch
+       * is started.
+       */
+      const reset = () => {
+        commands.push(currentCommand);
+        namespaces.clear();
+        currentNamespaceIndex = 0;
+        currentCommand = this.baseCommand();
+        namespaces.set(ns, currentNamespaceIndex);
+      };
+
       if (index != null) {
-        operations.push(buildOperation(model, index, this.pkFactory));
+        // Pushing to the ops document sequence returns the bytes length added.
+        const operation = buildOperation(model, index, this.pkFactory);
+        const operationBuffer = BSON.serialize(operation);
+
+        // Check if the operation buffer can fit in the current command. If it can,
+        // then add the operation to the document sequence and increment the
+        // current length as long as the ops don't exceed the maxWriteBatchSize.
+        if (
+          currentCommandLength + operationBuffer.length < maxMessageSizeBytes &&
+          currentCommand.ops.documents.length < maxWriteBatchSize
+        ) {
+          // Pushing to the ops document sequence returns the bytes length added.
```

Comment:
```suggestion
          // Pushing to the ops document sequence returns the total byte length of the document sequence.
```

Reply:
(this applies to other places in this file too)

Reply:
Updated.